File: samples/sample.hs
module Main where

import Data.List (sort)

-- Type alias
type Foo = String
type Bar = Int

-- Data type
data Baz = Baz Foo Bar

-- Function
foobar :: Foo -> Bar -> Baz
foobar foo bar = Baz foo bar

-- Pattern matching
bazValue :: Baz -> Foo
bazValue (Baz foo _) = foo

-- List comprehension
listComprehension :: [Bar] -> [Bar]
listComprehension xs = [x * 2 | x <- xs, x * 2 > 10]

-- Higher order function
applyTwice :: (a -> a) -> a -> a
applyTwice f x = f (f x)

-- IO action
main :: IO ()
main = do
  putStrLn "Hello, world!"
  let foo = "Lorem Ipsum"
  let bar = 123
  let baz = foobar foo bar
  putStrLn $ bazValue baz
  let xs = [1..10]
  print $ listComprehension xs
  print $ applyTwice (*2) bar

File: samples/sample.clj
(ns foobar.core
  (:require [clojure.string :as str]))

;; Defining a variable
(def foo "Lorem Ipsum")

;; Defining a function
(defn bar [x]
  (* x 2))

;; Using the function
(println (bar 123))

;; Using a list
(def baz [1 2 3 4 5])
(doseq [i baz]
  (println i))

;; Using a map
(def qux {"foo" "bar", "baz" "qux"})
(doseq [[k v] qux]
  (println (str "Key: " k ", Value: " v)))

;; Using control structures
(if (contains? qux "foo")
  (println (str "Found foo: " (get qux "foo")))
  (println "Did not find foo"))

;; Using a loop and recur
(loop [i 5]
  (when (> i 0)
    (println i)
    (recur (dec i))))

;; Using a higher-order function
(defn apply-to-foobar [f]
  (f foo))

(println (apply-to-foobar str/upper-case))
